src/builtins.rs

```diff
@@ -0,0 +1,686 @@
+                NixValue::Str(ref x) => x.clone(),
+                x => return Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+            };
+            new_map.remove(&key);
+        }
+        Ok(Gc::new(NixValue::Map(new_map)))
+    }
+    "intersectAttrs" ; IntersectAttrs => |param: Gc<Tree>| {
+        Ok(Gc::new(NixValue::Lambda(NixLambda::Builtin(NixBuiltin::IntersectAttrs1(param)))))
+    }
+    "intersectAttrs <?>" ; IntersectAttrs1(_0: Gc<Tree>) => |param: Gc<Tree>, a: &Gc<Tree>| {
+        let a = match a.eval()?.borrow() {
+            NixValue::Map(x) => x.clone(),
+            x => return Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+        };
+        let b = match param.eval()?.borrow() {
+            NixValue::Map(x) => x.clone(),
+            x => return Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+        };
+        let mut map = HashMap::new();
+        for (bkey, bval) in b.iter() {
+            if a.contains_key(bkey) {
+                map.insert(bkey.clone(), bval.clone());
+            }
+        }
+        Ok(Gc::new(NixValue::Map(map)))
+    }
+    "concatLists" ; ConcatLists => |param: Gc<Tree>| {
+        let list = match param.eval()?.borrow() {
+            NixValue::List(ref x) => x.clone(),
+            x => return Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+        };
+        let mut out = vec![];
+        for item in list {
+            let sublist = match item.eval()?.borrow() {
+                NixValue::List(ref x) => x.clone(),
+                x => return Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+            };
+            out.extend(sublist);
+        }
+        Ok(Gc::new(NixValue::List(out)))
+    }
+    "substring" ; Substring => |param: Gc<Tree>| {
+        Ok(Gc::new(NixValue::Lambda(NixLambda::Builtin(NixBuiltin::Substring1(match param.eval()?.borrow() {
+            NixValue::Integer(x) => *x as usize,
+            x => return Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+        })))))
+    }
+    "substring <?>" ; Substring1(_0: usize) => |param: Gc<Tree>, from: &usize| {
+        Ok(Gc::new(NixValue::Lambda(NixLambda::Builtin(NixBuiltin::Substring2(*from, match param.eval()?.borrow() {
+            NixValue::Integer(x) => *x as usize,
+            x => return Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+        })))))
+    }
+    "substring <?> <?>" ; Substring2(_0: usize, _1: usize) => |param: Gc<Tree>, from: &usize, len: &usize| {
+        let param = param.eval()?.as_str()?;
+        Ok(Gc::new(NixValue::Str(param[*from..(*from + *len).min(param.len())].to_string())))
+    }
+    "compareVersions" ; CompareVersions => |param| {
+        Ok(Gc::new(NixValue::Lambda(NixLambda::Builtin(NixBuiltin::CompareVersions1(param)))))
+    }
+    "compareVersions <?>" ; CompareVersions1(_0: Gc<Tree>) => |param: Gc<Tree>, left: &Gc<Tree>| {
```

Comment:
Can't we use some kind of semver crate in here?

Reply:
Perhaps, although we'll want to make sure we're using the exact same logic as official Nix. I'll make sure to link to the [relevant section in the Nix manual](https://nixos.org/manual/nix/stable/#ssec-version-comparisons) in code comments.

Reply:
semver crates would be likely not bug-compatible with nix itself.